File: backend/app/api/accommodation/guest_form.py
"""What a booking asks of its guests, and whether the answers are acceptable.

Two questions, deliberately answered in two different places:

    the step   -> *what* is asked (``template_config.guest_form``)
    a property -> *whether it is asked here* (``guest_form_mode``, and its
                  own form when it overrides)

The step owns one form for the whole checkout so an operator configures it
once. A property that needs something different (the hotel that wants a
passport number when the campsite does not) overrides it; a property that
should ask nothing turns it off. ``inherit`` is the default so a step's form
works the moment it is saved, without visiting every property.

Inside a form there are two sections. ``booker`` is asked once per room, of
whoever the room is for. ``guests`` is asked once per additional occupant,
and its ``mode`` is what keeps the common case cheap: ``same_as_booker``
repeats the booker's fields without duplicating the list, ``custom`` asks
something shorter, ``off`` asks nothing.

The guest's *name* is not a field here. It is a column on the booking, it is
what the property owner needs for their own registry, and it is governed by
the step's ``require_guest_names``. Making it configurable would mean a form
could be saved that leaves a property unable to tell one guest from another.
"""

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, model_validator

#: Field types a guest form may use. A subset of ``FormFieldType``: what is
#: missing is missing on purpose. ``signature`` and ``image_upload`` need an
#: upload target and a viewer that does not exist on this path yet, and
#: ``select_cards`` / ``multiselect_detailed`` are merchandising controls, not
#: questions. ``rich_text`` earns its place as the consent checkbox.
GUEST_FORM_FIELD_TYPES: frozenset[str] = frozenset(
    {
        "text",
        "textarea",
        "number",
        "boolean",
        "select",
        "radio",
        "multiselect",
        "date",
        "email",
        "url",
        "phone",
        "country_select",
        "rich_text",
    }
)

#: Types whose answer must be one of ``options``.
_CHOICE_TYPES = frozenset({"select", "radio"})

#: The most fields one section may carry. Not a technical limit: a checkout
#: that asks thirty questions per guest does not get filled in, and a typo in
#: an import should not become a wall of inputs.
MAX_FIELDS_PER_SECTION = 30

GuestFormMode = Literal["inherit", "off", "custom"]


class GuestFormField(BaseModel):
    """One question.

    Mirrors ``FormFieldSchema`` in ``@edgeos/shared-form-ui`` so the portal
    renders it with the same component as every other form, plus ``key``:
    the slug the answer is stored under. The key is generated from the label
    when the field is created and then frozen, so renaming a question does
    not orphan the answers already collected under it.
    """

    key: str
    type: str = "text"
    label: str
    required: bool = False
    placeholder: str | None = None
    help_text: str | None = None
    options: list[str] = []
    width: Literal["full", "half", "half_row"] | None = None
    config: dict[str, Any] = {}

    model_config = ConfigDict(str_strip_whitespace=True)

    @model_validator(mode="after")
    def _validate(self) -> "GuestFormField":
        if not self.key:
            raise ValueError("guest form field key cannot be empty")
        if self.type not in GUEST_FORM_FIELD_TYPES:
            allowed = ", ".join(sorted(GUEST_FORM_FIELD_TYPES))
            raise ValueError(
                f"guest form field type {self.type!r} is not allowed; use one of: {allowed}"
            )
        if not self.label:
            raise ValueError(f"guest form field {self.key!r} needs a label")
        if self.type in _CHOICE_TYPES and not self.options:
            raise ValueError(f"guest form field {self.key!r} needs options")
        return self


class GuestFormSection(BaseModel):
    """A block of questions with a heading."""

    title: str | None = None
    description: str | None = None
    fields: list[GuestFormField] = []

    model_config = ConfigDict(str_strip_whitespace=True)

    @model_validator(mode="after")
    def _validate(self) -> "GuestFormSection":
        keys = [field.key for field in self.fields]
        if len(keys) != len(set(keys)):
            raise ValueError("guest form field keys must be unique within a section")
        if len(keys) > MAX_FIELDS_PER_SECTION:
            raise ValueError(
                f"a guest form section may hold at most {MAX_FIELDS_PER_SECTION} fields"
            )
        return self


class GuestFormGuestsSection(GuestFormSection):
    """The questions asked of each additional occupant.

    ``same_as_booker`` is the default because it is the common case and the
    one that stays right when the booker's fields change later.
    """

    mode: Literal["same_as_booker", "custom", "off"] = "same_as_booker"


class AccommodationGuestForm(BaseModel):
    """The whole form: what the booker is asked, and what each guest is asked."""

    version: int = 1
    booker: GuestFormSection = GuestFormSection()
    guests: GuestFormGuestsSection = GuestFormGuestsSection()

    model_config = ConfigDict(extra="allow")

    def guest_fields(self) -> list[GuestFormField]:
        """The fields each additional guest is asked, ``mode`` resolved."""
        if self.guests.mode == "off":
            return []
        if self.guests.mode == "same_as_booker":
            return list(self.booker.fields)
        return list(self.guests.fields)

    def asks_nothing(self) -> bool:
        return not self.booker.fields and not self.guest_fields()


class GuestFormError(Exception):
    """An answer the form does not accept.

    Carries the offending field's ``key`` so the checkout can put the message
    next to the input instead of at the bottom of the page. ``guest_index`` is
    ``None`` for the booker and the zero-based occupant otherwise.
    """

    def __init__(self, message: str, *, key: str, guest_index: int | None = None):
        super().__init__(message)
        self.message = message
        self.key = key
        self.guest_index = guest_index


def parse_form(raw: Any) -> AccommodationGuestForm | None:
    """Read a stored form, or ``None`` when there is nothing to ask.

    Never raises: this reads what is already in the database, and a form that
    became unparseable (an older shape, a hand-edited row) must not take the
    checkout down with it. Writes are validated on the way in, which is where
    a bad form should be rejected.
    """
    if not raw:
        return None
    try:
        form = AccommodationGuestForm.model_validate(raw)
    except Exception:
        return None
    return None if form.asks_nothing() else form


def resolve_form(
    step_form: Any,
    property_row: Any,
) -> AccommodationGuestForm | None:
    """The form actually asked for a booking at this property.

    ``step_form`` is the step's ``template_config.guest_form``, as stored.

    ``off`` wins over anything the step configured, and ``custom`` replaces it
    rather than adding to it: an override that silently kept inheriting half
    the questions would be impossible to reason about from the property page.
    """
    mode = getattr(property_row, "guest_form_mode", None) or "inherit"

    if mode == "off":
        return None
    if mode == "custom":
        return parse_form(getattr(property_row, "guest_form", None))
    return parse_form(step_form)


def snapshot(form: AccommodationGuestForm | None) -> dict[str, Any] | None:
    """The form as it stood when the booking was made.

    Stored on the booking so the export and the booking detail can label an
    answer with the question that was actually asked. Without it, renaming a
    field would relabel every answer ever given, including the ones already
    sent to a property owner.
    """
    return form.model_dump(mode="json") if form else None


def _is_blank(value: Any) -> bool:
    if value is None:
        return True
    if isinstance(value, str):
        return not value.strip()
    if isinstance(value, list | dict):
        return not value
    return False


def _check_field(
    field: GuestFormField,
    answers: dict[str, Any],
    guest_index: int | None,
) -> None:
    value = answers.get(field.key)

    def fail(message: str) -> GuestFormError:
        return GuestFormError(message, key=field.key, guest_index=guest_index)

    if _is_blank(value):
        # `boolean` is the exception that proves the rule: a required consent
        # checkbox is answered by being ticked, and False is an answer that
        # does not satisfy it.
        if field.required:
            raise fail(f"{field.label} is required")
        return

    if field.type == "boolean":
        if field.required and value is not True:
            raise fail(f"{field.label} is required")
        return

    if field.type in _CHOICE_TYPES and value not in field.options:
        raise fail(f"{field.label}: pick one of the offered options")

    if field.type == "multiselect":
        values = value if isinstance(value, list) else [value]
        if field.options and any(item not in field.options for item in values):
            raise fail(f"{field.label}: pick from the offered options")

    if field.type == "email" and not _looks_like_email(value):
        raise fail(f"{field.label} does not look like an email address")

    if field.type == "number":
        number = _as_number(value)
        if number is None:
            raise fail(f"{field.label} must be a number")
        minimum = field.config.get("min")
        maximum = field.config.get("max")
        if minimum is not None and number < float(minimum):
            raise fail(f"{field.label} must be {minimum} or more")
        if maximum is not None and number > float(maximum):
            raise fail(f"{field.label} must be {maximum} or less")


def _looks_like_email(value: Any) -> bool:
    text = str(value).strip()
    if " " in text or text.count("@") != 1:
        return False
    local, _, domain = text.partition("@")
    return bool(local) and "." in domain and not domain.startswith(".")


def _as_number(value: Any) -> float | None:
    try:
        return float(str(value).strip())
    except (TypeError, ValueError):
        return None


def validate_answers(
    form: AccommodationGuestForm | None,
    booker_answers: dict[str, Any] | None,
    guests: list[dict[str, Any]] | None,
) -> None:
    """Raise :class:`GuestFormError` on the first answer the form refuses.

    The booker is checked before the guests, and each guest in order, so the
    message always points at the first input the buyer would reach going down
    the page. Answers for fields the form does not define are ignored rather
    than rejected: a form that lost a question should not make an in-flight
    cart unpayable.
    """
    if form is None:
        return

    for field in form.booker.fields:
        _check_field(field, booker_answers or {}, None)

    guest_fields = form.guest_fields()
    if not guest_fields:
        return

    for index, guest in enumerate(guests or []):
        answers = guest.get("answers") if isinstance(guest, dict) else None
        for field in guest_fields:
            _check_field(field, answers or {}, index)


def normalise_guests(raw: Any) -> list[dict[str, Any]]:
    """Guests as they are stored: ``{"name": ..., "answers": {...}}``.

    Accepts the shape carts and clients used before this existed, where a
    guest was a bare name, so a cart saved yesterday still purchases today.
    Entries with neither a name nor an answer are dropped: they are the empty
    slots the checkout renders for a party that has not been filled in.
    """
    out: list[dict[str, Any]] = []
    for entry in raw or []:
        if isinstance(entry, str):
            name, answers = entry.strip(), {}
        elif isinstance(entry, dict):
            name = str(entry.get("name") or "").strip()
            raw_answers = entry.get("answers")
            answers = raw_answers if isinstance(raw_answers, dict) else {}
        else:
            continue
        if not name and not answers:
            continue
        out.append({"name": name, "answers": answers})
    return out

File: backoffice/src/components/accommodations/guest-form/guestForm.ts
/**
 * The questions this checkout asks about the people staying.
 *
 * Mirrors `backend/app/api/accommodation/guest_form.py`: the same field
 * types, the same two sections, the same guests mode. What the backend
 * refuses on write, this refuses to build, so an operator does not discover
 * a bad form by way of a 422.
 *
 * A field's `key` is the slug the answer is stored under. It is generated
 * from the label once, at creation, and then frozen: renaming a question
 * later must not orphan the answers already collected under the old key.
 */

import type { FormFieldSchema } from "@edgeos/shared-form-ui"
import { FIELD_TYPES, slugify } from "@/components/form-builder"

/** The subset of field types a guest form may use.
 *
 * Kept in step with `GUEST_FORM_FIELD_TYPES` on the backend. What is missing
 * is missing on purpose: signature and image upload need an upload target
 * this path does not have, and the card pickers are merchandising controls
 * rather than questions.
 */
export const GUEST_FIELD_TYPES = FIELD_TYPES.filter((type) =>
  [
    "text",
    "textarea",
    "number",
    "boolean",
    "select",
    "radio",
    "multiselect",
    "date",
    "email",
    "url",
    "phone",
    "country_select",
    "rich_text",
  ].includes(type.value),
)

/** Types whose answer is picked from `options`, so the field needs some. */
export const CHOICE_TYPES = new Set(["select", "radio", "multiselect"])

/** Matches `MAX_FIELDS_PER_SECTION` on the backend. */
export const MAX_FIELDS = 30

/** Past this, the editor says so. Not a limit, a nudge: a checkout that asks
 *  ten questions per guest does not get filled in. */
export const CROWDED_AFTER = 6

export type GuestsMode = "same_as_booker" | "custom" | "off"

export interface GuestField {
  key: string
  type: string
  label: string
  required: boolean
  placeholder?: string | null
  help_text?: string | null
  options: string[]
  width?: "full" | "half" | "half_row" | null
  config: Record<string, unknown>
}

export interface GuestFormValue {
  version: number
  booker: {
    title?: string | null
    description?: string | null
    fields: GuestField[]
  }
  guests: {
    mode: GuestsMode
    title?: string | null
    description?: string | null
    fields: GuestField[]
  }
}

export function emptyForm(): GuestFormValue {
  return {
    version: 1,
    booker: { title: null, description: null, fields: [] },
    guests: {
      mode: "same_as_booker",
      title: null,
      description: null,
      fields: [],
    },
  }
}

function parseField(raw: unknown): GuestField | null {
  if (!raw || typeof raw !== "object") return null
  const field = raw as Record<string, unknown>
  if (typeof field.key !== "string" || !field.key) return null
  return {
    key: field.key,
    type: typeof field.type === "string" ? field.type : "text",
    label: typeof field.label === "string" ? field.label : field.key,
    required: field.required === true,
    placeholder:
      typeof field.placeholder === "string" ? field.placeholder : null,
    help_text: typeof field.help_text === "string" ? field.help_text : null,
    options: Array.isArray(field.options)
      ? field.options.filter(
          (option): option is string => typeof option === "string",
        )
      : [],
    width: (field.width as GuestField["width"]) ?? null,
    config:
      field.config && typeof field.config === "object"
        ? (field.config as Record<string, unknown>)
        : {},
  }
}

/** Read a stored form. Forgiving, like the backend's reader: a shape that no
 *  longer makes sense should leave the editor empty, not blank the screen. */
export function parseForm(raw: unknown): GuestFormValue {
  if (!raw || typeof raw !== "object") return emptyForm()
  const form = raw as Record<string, unknown>
  const booker = (form.booker ?? {}) as Record<string, unknown>
  const guests = (form.guests ?? {}) as Record<string, unknown>
  const mode = guests.mode
  return {
    version: typeof form.version === "number" ? form.version : 1,
    booker: {
      title: (booker.title as string) ?? null,
      description: (booker.description as string) ?? null,
      fields: (Array.isArray(booker.fields) ? booker.fields : [])
        .map(parseField)
        .filter((field): field is GuestField => field !== null),
    },
    guests: {
      mode:
        mode === "custom" || mode === "off" || mode === "same_as_booker"
          ? mode
          : "same_as_booker",
      title: (guests.title as string) ?? null,
      description: (guests.description as string) ?? null,
      fields: (Array.isArray(guests.fields) ? guests.fields : [])
        .map(parseField)
        .filter((field): field is GuestField => field !== null),
    },
  }
}

/** The fields each additional guest is asked, `mode` resolved. */
export function guestFieldsOf(form: GuestFormValue): GuestField[] {
  if (form.guests.mode === "off") return []
  if (form.guests.mode === "same_as_booker") return form.booker.fields
  return form.guests.fields
}

/** A form that asks nothing is no form: it is stored as null, so the
 *  checkout asks nothing rather than rendering an empty panel. */
export function isEmpty(form: GuestFormValue | null | undefined): boolean {
  if (!form) return true
  return form.booker.fields.length === 0 && guestFieldsOf(form).length === 0
}

/** The form as `template_config.guest_form` holds it.
 *
 * Untyped on the wire, because `template_config` is a plain JSON column and
 * the generated client has no shape for what lives inside it. The backend
 * validates it on write, which is where a bad form should be caught. */
export function toApi(
  form: GuestFormValue | null | undefined,
): GuestFormValue | null {
  return isEmpty(form) || !form ? null : form
}

/** A key nothing else in this form is using.
 *
 * Falls back to the type when the label slugifies to nothing (a label of
 * "?" or of non-latin script), because a field with no key cannot store an
 * answer at all.
 */
export function makeKey(label: string, type: string, taken: string[]): string {
  const base = slugify(label) || slugify(type) || "field"
  if (!taken.includes(base)) return base
  let suffix = 2
  while (taken.includes(`${base}_${suffix}`)) suffix += 1
  return `${base}_${suffix}`
}

export function makeField(
  type: string,
  label: string,
  taken: string[],
): GuestField {
  return {
    key: makeKey(label, type, taken),
    type,
    label,
    required: false,
    placeholder: null,
    help_text: null,
    options: CHOICE_TYPES.has(type) ? ["Option 1", "Option 2"] : [],
    width: null,
    config: {},
  }
}

/** Every key the form already uses, both sections, so a new one is unique
 *  across the whole thing and copying a field between sections stays safe. */
export function takenKeys(form: GuestFormValue): string[] {
  return [...form.booker.fields, ...form.guests.fields].map(
    (field) => field.key,
  )
}

/** What a field is missing before it can be saved. Mirrors the backend's
 *  write validation, so nothing reaches the API only to bounce. */
export function fieldProblem(field: GuestField): string | null {
  if (!field.label.trim()) return "This question needs a label"
  if (CHOICE_TYPES.has(field.type) && field.options.length === 0) {
    return "Add at least one option to pick from"
  }
  return null
}

export function formProblem(form: GuestFormValue): string | null {
  for (const field of [...form.booker.fields, ...form.guests.fields]) {
    const problem = fieldProblem(field)
    if (problem) return `${field.label || "A question"}: ${problem}`
  }
  if (form.booker.fields.length > MAX_FIELDS) {
    return `The booking contact section can hold at most ${MAX_FIELDS} questions`
  }
  if (form.guests.fields.length > MAX_FIELDS) {
    return `The guests section can hold at most ${MAX_FIELDS} questions`
  }
  return null
}

/** The shape `SchemaField` renders, so the preview is the real control and
 *  not a drawing of one. */
export function toSchemaField(field: GuestField): FormFieldSchema {
  return {
    type: field.type as FormFieldSchema["type"],
    label: field.label,
    required: field.required,
    options: field.options.length ? field.options : undefined,
    placeholder: field.placeholder ?? undefined,
    help_text: field.help_text ?? undefined,
    width: field.width ?? null,
    config: field.config,
  }
}

/**
 * Questions the checkout answers by itself, having already asked the buyer.
 *
 * Mirrors `COVERABLE` in the portal's `lib/buyerIdentity.ts`, and matches on
 * the key for the same reason: a key is generated once and frozen, so
 * `email` means the buyer's email in every form on every tenant, while a
 * second field that merely has type `email` is usually somebody else's
 * address and must go on being asked.
 *
 * `always` holds everywhere: a direct sale's buyer step forces email and
 * both name halves into its schema, and an application flow reads the same
 * three off the signed-in account. `if_asked` depends on the popup's own
 * buyer form, which this editor does not load, so it is stated as the
 * condition it is rather than guessed at.
 */
export type BuyerCoverage = "always" | "if_asked"

const BUYER_COVERAGE: Record<string, BuyerCoverage> = {
  email: "always",
  first_name: "always",
  last_name: "always",
  name: "always",
  full_name: "always",
  phone: "if_asked",
}

/** How the checkout covers this question, or null when it is the form's own. */
export function buyerCoverage(field: GuestField): BuyerCoverage | null {
  return BUYER_COVERAGE[field.key] ?? null
}

// ---------------------------------------------------------------------------
// Presets
// ---------------------------------------------------------------------------

function build(
  fields: Array<[string, string, boolean]>,
  guests: GuestsMode = "same_as_booker",
): GuestFormValue {
  const form = emptyForm()
  form.guests.mode = guests
  for (const [type, label, required] of fields) {
    const field = makeField(type, label, takenKeys(form))
    field.required = required
    form.booker.fields.push(field)
  }
  return form
}

export interface GuestFormPreset {
  key: string
  label: string
  description: string
  build: () => GuestFormValue
}

/** Starting points, because a blank editor is the slowest way to begin.
 *
 * Every preset asks the *booking contact* the questions and repeats them for the
 * others; narrowing that is one click in the guests section.
 *
 * None of them asks for a name or an email any more. The checkout already
 * has both, from the buyer step or from the signed-in account, and fills
 * them into the booking on the way to the property. A preset that asked for
 * them would be building the duplication these were rewritten to remove.
 */
export const PRESETS: GuestFormPreset[] = [
  {
    key: "contact",
    label: "A phone number",
    description:
      "The one contact detail the checkout does not already have from the buyer",
    build: () => build([["phone", "Phone", false]]),
  },
  {
    key: "registry",
    label: "Hotel registry",
    description:
      "What a front desk is usually required to record, beyond the name and email",
    build: () =>
      build([
        ["phone", "Phone", true],
        ["date", "Date of birth", true],
        ["country_select", "Nationality", true],
        ["text", "ID or passport number", true],
      ]),
  },
  {
    key: "stay_needs",
    label: "Needs for the stay",
    description: "What the property should prepare for, asked of everyone",
    build: () =>
      build([
        ["textarea", "Dietary requirements", false],
        ["textarea", "Accessibility needs", false],
      ]),
  },
  {
    key: "blank",
    label: "Start from scratch",
    description: "An empty form to add your own questions to",
    build: emptyForm,
  },
]
